### ISSUE TITLE
Missing Response on Shell Channel

### ISSUE THREAD

**Issue by Artjom-Metro** [alexarchambault/jupyter-scala#66]
Hi,

I am right now trying to develop an alternative Jupyter kernel manager, but during testing I had the problem when I worked with your kernel that I didn't get a response to my messages on the the shell channel.

I manually start the your kernel like this:

```Bash
$ java -jar ~/.ipython/kernels/scala211/launcher.jar --connection-file  /tmp/scala_kernel/connection.json
```

Here the connection file:

```Json
{
  "stdin_port": 53888, 
  "ip": "127.0.0.1", 
  "control_port": 53889, 
  "hb_port": 53890, 
  "signature_scheme": "hmac-sha256", 
  "key": "", 
  "kernel_name": "", 
  "shell_port": 53886, 
  "transport": "tcp", 
  "iopub_port": 53887
}
```

And then run the following Scala code:

```Scala
import org.zeromq.{ZMQ, ZMsg}
import scala.collection.JavaConversions._
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global

/**
  * Created by artjom on 14.04.16.
  */
object ScalaKernelProblem extends App {


  val msgTxt =
    s"""<IDS|MSG>
        |6ea6b213262402cc1ad3c1d3e342a9f6
        |{"date":"2013-04-27T23:22:13.522049","username":"test","session":"examplesession","msg_id":"examplemessage","msg_type":"execute_request"}
        |{}
        |{}
        |{"code":"68 - 26","silent":false,"allow_stdin":true,"store_history":true,"user_expressions":{}, "stop_on_error": false}
      """.stripMargin

  val msg = new ZMsg()
  msgTxt.split("\n").filter(!_.isEmpty).foreach(msg.add)


  val context = ZMQ.context(1)

  val iopub = context.socket(ZMQ.SUB)
  iopub.subscribe(Array.empty[Byte])
  iopub.connect("tcp://127.0.0.1:53887")

  val socket = context.socket(ZMQ.REQ)
  socket.connect("tcp://127.0.0.1:53886")

  println("Listening on iopub channel")
  Future {
      while(true){
        Option(ZMsg.recvMsg(iopub, 0)).foreach(zMsg =>
        zMsg.toSeq.map(zFrame => new String(zFrame.getData, ZMQ.CHARSET)).foreach(s => println(s"iopub> $s"))
        )
        println("Further listening")
      }
    }



  println("Sending msg...")
  val b = msg.send(socket)
  println(s"Send message ($b). Waiting for response...")
  val reply = ZMsg.recvMsg(socket, 0).toList
  println(s"Got response: $reply. Finished")

  Thread.sleep(1000)

}

```

The output looks like this:

```Bash
> !!
kernel_manager/test:runMain com.test.kernel_manager.ScalaKernelProblem
[info] Running com.test.kernel_manager.ScalaKernelProblem 
[info] Listening on iopub channel
[info] Sending msg...
[info] Send message (true). Waiting for response...
[info] iopub> execute_input
[info] iopub> <IDS|MSG>
[info] iopub> 
[info] iopub> {"msg_id":"b2aa05e0-4533-4ac9-b777-ef07ad63da06","username":"test","msg_type":"execute_input","version":null,"session":"examplesession"}
[info] iopub> {"msg_id":"examplemessage","username":"test","msg_type":"execute_request","version":null,"session":"examplesession"}
[info] iopub> {}
[info] iopub> {"execution_count":1,"code":"68 - 26"}
[info] Further listening
[info] iopub> status
[info] iopub> <IDS|MSG>
[info] iopub> 
[info] iopub> {"msg_id":"3b077df4-c16e-4692-8cf4-e6c2d97d16dd","username":"test","msg_type":"status","version":"5.0","session":"examplesession"}
[info] iopub> {"msg_id":"examplemessage","username":"test","msg_type":"execute_request","version":null,"session":"examplesession"}
[info] iopub> {}
[info] iopub> {"execution_state":"busy"}
[info] Further listening
[info] iopub> display_data
[info] iopub> <IDS|MSG>
[info] iopub> 
[info] iopub> {"msg_id":"3099b7f9-2bd0-4e19-b3b2-5fbbd0b66f96","username":"test","msg_type":"display_data","version":null,"session":"examplesession"}
[info] iopub> {"msg_id":"examplemessage","username":"test","msg_type":"execute_request","version":null,"session":"examplesession"}
[info] iopub> {}
[info] iopub> {"metadata":{},"data":{"text/plain":"\u001b[36mres0\u001b[0m: \u001b[32mInt\u001b[0m = \u001b[32m42\u001b[0m"},"source":"interpreter"}
[info] Further listening
[info] iopub> status
[info] iopub> <IDS|MSG>
[info] iopub> 
[info] iopub> {"msg_id":"cb3252c7-c25b-4192-8567-be56a1d0518f","username":"test","msg_type":"status","version":"5.0","session":"examplesession"}
[info] iopub> {"msg_id":"examplemessage","username":"test","msg_type":"execute_request","version":null,"session":"examplesession"}
[info] iopub> {}
[info] iopub> {"execution_state":"idle"}
[info] Further listening
```

---

So, as you can see I get the expected messages on the iopub channel, but the programs hangs forever because I actually never get the response on the shell channel that the code was executed. The jupyter documentation describes this step [here](http://jupyter-client.readthedocs.org/en/latest/messaging.html#messages-on-the-shell-router-dealer-sockets):

>Upon completion of the execution request, the kernel always sends a reply, with a status code indicating what happened and additional data depending on the outcome. See below for the possible return codes and associated data.

So my question basically is: Is there any way to force a response on the shell channel (e.g. change my request method) or if it is a bug, is there a more recent version where the problem is fixed?

Best regards,
Artjom

---
I am using Scala 2.11.7, version 0.3.5 of jeromq and  kernel_2.11-0.3.0-M2.jar of your kernel (at least that is what I can find in the `~/.jupyter-scala/bootstrap`)

**Comment by alexarchambault** [comment]
It should send something on the shell channel, it's done [here](https://github.com/alexarchambault/jupyter-kernel/blob/5d3f344812ec9b967ff689fc1794394f2f338d4e/core/src/main/scala/jupyter/kernel/interpreter/InterpreterHandler.scala#L72).

Are you sure the problem couldn't come from the ZMQ setup? Like `socket` being a `ZMQ.REQ`. `shell` channels should be router / dealer ones I think (so says [your link](http://jupyter-client.readthedocs.org/en/latest/messaging.html#messages-on-the-shell-router-dealer-sockets)). They're setup [this way](https://github.com/alexarchambault/jupyter-kernel/blob/5d3f344812ec9b967ff689fc1794394f2f338d4e/core/src/main/scala/jupyter/kernel/stream/zmq/ZMQStreams.scala#L33) in jupyter-scala.

**Comment by Artjom-Metro** [comment]
Thanks for your quick response. 
After some hours of debugging I finally found the problem, you are right that your application was preparing to send the execution confirmation, but it was actually never send. The problem there is that the ZMQ Router silently drops messages if it has no identifier information where to send it. [Here] is the relevant part of the documentation, I could easily solve this problem by setting an identifier explicitly in my code:

```Scala
val socket = context.socket(ZMQ.REQ)
  socket.setIdentity("manager_id".getBytes(ZMQ.CHARSET))
  socket.connect("tcp://127.0.0.1:53886")
```

So this issue is actually my fault (as you expected)

---

But one problem I encountered while debugging this issue is that in your [message receiving method](https://github.com/alexarchambault/jupyter-kernel/blob/5d3f344812ec9b967ff689fc1794394f2f338d4e/core/src/main/scala/jupyter/kernel/stream/zmq/ZMQStreams.scala#L147), you require the user to use a non-empty key. Otherwise the code will try to use the identifier, which is the first part of the message, as the header, because the check for the delimiter is ignored. Although the workaround is of course easy, if you know the problem is, I think this could be fixed without much effort.
